stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -2371,20 +2371,76 @@ buildEnvironmentPath(
   return totalKeyParamCount;
 }
 
+unsigned SubstGenericParametersFromMetadata::buildShapePath(
+    const TargetExtendedExistentialTypeShape<InProcess> *shape) const {
+  unsigned totalParamCount = 0;
+  unsigned totalKeyParamCount = 0;
+
+  auto genSig = shape->getGeneralizationSignature();
+  if (!genSig.getParams().empty()) {
+    bool hasNonKeyGenericParams = false;
+    unsigned numKeyGenericParamsHere = 0;
+    for (const auto &gp : genSig.getParams()) {
+      if (gp.hasKeyArgument())
```

Comment:
The generalization and requirement signatures should only have key arguments; I think we assert that elsewhere.  So this should be straightforward indexing.